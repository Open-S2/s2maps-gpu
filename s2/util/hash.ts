import type { View } from 'ui/camera/projector/index.js';

/**
 * Sets the hash in the location bar given the current view
 * @param view - The current camera view
 */
export function setHash(view: Required<View>): void {
  const { bearing, pitch } = view;
  let { lon, lat, zoom } = view;
  zoom = Math.round(zoom * 100) / 100;
  // derived from equation: 512px * 2^z / 360 / 10^d < 0.5px
  const precision = Math.ceil((zoom * Math.LN2 + Math.log(512 / 360 / 0.5)) / Math.LN10);
  const m = Math.pow(10, precision);
  lon = Math.round(lon * m) / m;
  lat = Math.round(lat * m) / m;
  let hash = `#${zoom}/${lon}/${lat}`;

  if (bearing !== 0) hash += `/${Math.round(bearing * 10) / 10}`;
  if (pitch !== 0) hash += `/${Math.round(pitch)}`;

  // set hash on the location bar
  const location = window.location.href.replace(/(#.+)?$/, hash);
  window.history.replaceState(window.history.state, '', location);
}

/** @returns The current view converted from the hash */
export function parseHash(): View {
  const view: View = {};
  const [, hash] = window.location.href.split('#');
  // corner case: no hash
  if (hash === undefined) return view;
  // parse hash
  const [zoom, lon, lat, bearing, pitch] = hash.split('/') as [
    zoom: string,
    lon: string,
    lat: string,
    bearing: string | undefined,
    pitch: string | undefined,
  ];
  view.zoom = parseFloat(zoom);
  view.lon = parseFloat(lon);
  view.lat = parseFloat(lat);
  if (bearing !== undefined) view.bearing = parseFloat(bearing);
  if (pitch !== undefined) view.pitch = parseFloat(pitch);

  return view;
}
