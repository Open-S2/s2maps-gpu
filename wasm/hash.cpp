// hash.cpp
typedef long int i32;
extern "C" {
  void hash(i32 *plaintext, i32 plaintextLength) {
    i32 hash;
    i32 i;
    for (hash = i = 0; i < plaintextLength; ++i) {
      hash += plaintext[i];
      hash += (hash << 10);
      hash ^= (hash >> 6);
    }
    hash += (hash << 3);
    hash ^= (hash >> 11);
    hash += (hash << 15);

    plaintext[0] = hash;
  }
}
