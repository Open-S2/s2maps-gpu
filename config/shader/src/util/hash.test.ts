import { toHash, toMurmur53 } from './hash';
import uniq from 'lodash/uniq';

const add = (a: number, b: number) => ((a|0) + (b|0)) >>> 0;
const rot = (a: number, b: number) => ((a << b) | (a >>> (32 - b))) >>> 0;
const mul = Math.imul;

describe('hash', () => {

  it("arithmetic", () => {
    expect(add(1, 1)).toEqual(2);
    expect(add(0x7fffffff, 1)).toEqual(0x80000000);
    expect(add(0xffffffff, 1)).toEqual(0);

    expect(mul(0x1000, 0x1000)).toEqual(0x1000000);
    expect(mul(0x7fffffff, 0x2)).toEqual(-2);
    expect(mul(0x80000001, 0x2)).toEqual(0x2);
    expect(mul(0x80000001, 0x80000001)).toEqual(0x1);

    expect(rot(0x7fffff00, 31)).toEqual(0x3fffff80);
    expect(rot(0x80000001, 31)).toEqual(0xc0000000);
  })

  it('hashes values', () => {

    const u81 = new Uint8Array(3);
    u81[0] = 1;
    u81[1] = 0;

    const u82 = new Uint8Array(3);
    u82[0] = 1;
    u82[1] = 2;

    const values = [
      undefined,
      null,
      "",
      "0",
      0,
      1,
      0x100000000,
      0x100000001,
      [],
      [0],
      [1],
      [1, 0],
      [1, 2],
      u81,
      u82,
      {},
      {f:0},
      {g:0},
      {f:1},
      {g:1},
    ];
    
    const hashes = values.map(toHash);
    
    expect(hashes).toMatchSnapshot();
    expect(uniq(hashes).length).toBe(hashes.length);
  });
  
  it("doesn't collide 16-bit ints", () => {

    const seen = new Map<number, number>();
    for (let i = 0; i < 0xffff; ++i) {
      const v = toMurmur53(i);
      if (seen.has(v)) {
        console.log('hash collision', seen.get(v), i, '=', v);
        expect(false).toBe(true);
      }
      seen.set(v, i);
    }

  });
});
