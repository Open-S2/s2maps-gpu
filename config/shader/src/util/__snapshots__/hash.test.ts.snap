// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hash hashes values 1`] = `
Array [
  "1l2jf6l28r",
  "8vthalggis",
  "uottwj0y3r",
  "3807eq7w5v",
  "cbnaeh3g55",
  "n09fwdb0u5",
  "5yz451x4f5",
  "a1mdu4pl9r",
  "pcbrt4hb9e",
  "6couupxn1k",
  "olq5wiy04o",
  "oa26q9aowj",
  "eikp8yldgd",
  "br94b82qgs",
  "ncsym7f7sa",
  "9q11fly507",
  "dpeqc1pdst",
  "bbkhj0cpxb",
  "tmmfq8wm90",
  "xavdphcmmm",
]
`;
